File: frontend/src/data/roles.js
// Libellé affichable d'un rôle (voir spec/SPEC.md §2, §4) — partagé entre
// Header (sélecteur famille) et ProfilScreen (liste des profils du foyer).
export const ROLE_LABEL = { admin: 'Admin', professeur: 'Professeur', eleve: 'Élève', superuser: 'Propriétaire' }

// Rang de rôle, du plus faible au plus fort (voir spec §2.2 "règle de
// sécurité du switch de profil famille") : sert à savoir si passer d'un
// profil à l'autre est une montée en privilège (code redemandé) ou non.
export const ROLE_RANK = { eleve: 0, professeur: 1, admin: 2, superuser: 3 }

// Passer de `depuisType` à `versType` est une montée en privilège (voir
// §2.2, règle Élève < Professeur < Admin) : le code d'accès du rôle visé
// doit alors être redemandé. Vers un rôle égal ou inférieur : bascule libre.
export function estMonteeEnPrivilege(depuisType, versType) {
  return ROLE_RANK[versType] > ROLE_RANK[depuisType]
}

// Ordre d'affichage du sélecteur de profil famille (Header) et de "Ma
// famille" (ProfilScreen) : Admin, puis Professeur, puis Élève — pas
// l'ordre arbitraire du tableau en mémoire.
export function trierParRole(profils) {
  return [...profils].sort((a, b) => ROLE_RANK[b.type] - ROLE_RANK[a.type])
}

// Rôles cumulables (voir spec/SPEC.md §2.1 et §6.3bis) : un profil peut en
// avoir plusieurs (ex. professeur ET admin), dans `profil.roles`.
// ⚠️ Pour savoir ce qu'un profil a le DROIT de faire, toujours passer par
// ces fonctions, jamais par `profil.type` : `type` n'est que le rôle
// principal (le plus élevé), bon pour un libellé ou un rang, mais un
// professeur-admin (type 'admin') y perdrait tout ce qui est propre aux
// profs (ex. "Mes heures" dans Profil). Même règle côté serveur
// (backend/src/comptes/roles.py).
function rolesDe(profil) {
  // Repli sur `type` pour un objet qui ne porte pas encore la liste (ex.
  // un membre de conversation) : il n'a alors qu'un seul rôle.
  const roles = profil?.roles ?? (profil?.type ? [profil.type] : [])
  // Le Superuser (§2.5) voit et fait tout ce que fait un Owner, dans
  // l'école qu'il a choisie : mêmes onglets, mêmes boutons. Le serveur
  // l'autorise de son côté (voir comptes/rbac.py).
  return roles.includes('superuser') ? [...roles, 'admin', 'owner'] : roles
}

export function aUnDesRoles(profil, roles) {
  return roles.some((role) => rolesDe(profil).includes(role))
}

export const isEleve = (profil) => aUnDesRoles(profil, ['eleve'])
export const isProf = (profil) => aUnDesRoles(profil, ['professeur'])
export const isAdmin = (profil) => aUnDesRoles(profil, ['admin'])
export const isOwner = (profil) => aUnDesRoles(profil, ['owner'])
export const isSuperuser = (profil) => aUnDesRoles(profil, ['superuser'])

File: backend/src/comptes/roles.py
"""Rôles cumulables d'un compte (voir spec/SPEC.md §2.1 et §6.3bis).

Un compte peut avoir plusieurs rôles (ex. professeur ET admin), stockés
dans `roles_compte` (voir models.py:RoleCompte). **Le reste du code ne lit
jamais ces lignes directement** : il passe par les fonctions ci-dessous.
Si un jour la façon de stocker les rôles change, seul ce fichier change.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Iterable

if TYPE_CHECKING:
    from .models import Compte

ELEVE = "eleve"
PROFESSEUR = "professeur"
ADMIN = "admin"
OWNER = "owner"
# Propriétaire de l'application, au-dessus des écoles (§2.5) : compte sans
# école, jamais cumulé, jamais attribué depuis l'appli.
SUPERUSER = "superuser"

ROLES = (ELEVE, PROFESSEUR, ADMIN, OWNER, SUPERUSER)

# Rang, du plus faible au plus fort (voir §2.2) : sert à savoir si passer
# d'un profil familial à l'autre est une montée en privilège, et à choisir
# le rôle principal affiché. Owner compte comme Admin : ce n'est qu'un
# droit en plus sur la gestion des admins (§2.4), pas un rang au-dessus.
# Même règle que le frontend (frontend/src/data/roles.js).
RANG = {ELEVE: 0, PROFESSEUR: 1, ADMIN: 2, OWNER: 2, SUPERUSER: 3}


def noms_roles(compte: Compte) -> list[str]:
    """Rôles du compte, du plus fort au plus faible — ordre stable pour
    l'API et les tests."""
    return sorted((r.role for r in compte.roles), key=lambda nom: (-RANG[nom], nom))


def a_le_role(compte: Compte, role: str) -> bool:
    return any(r.role == role for r in compte.roles)


def is_eleve(compte: Compte) -> bool:
    return a_le_role(compte, ELEVE)


def is_prof(compte: Compte) -> bool:
    return a_le_role(compte, PROFESSEUR)


def is_admin(compte: Compte) -> bool:
    return a_le_role(compte, ADMIN)


def is_owner(compte: Compte) -> bool:
    return a_le_role(compte, OWNER)


def is_superuser(compte: Compte) -> bool:
    return a_le_role(compte, SUPERUSER)


def role_principal(noms: Iterable[str]) -> str:
    """Le rôle le plus élevé, Owner ramené à Admin : c'est celui qu'on
    affiche comme libellé ("Admin", "Professeur"...) et qui sert de rang
    pour la bascule de profil familial (§2.2)."""
    noms = [ADMIN if nom == OWNER else nom for nom in noms]
    return max(noms, key=lambda nom: RANG[nom])


def rang(compte: Compte) -> int:
    return RANG[role_principal(r.role for r in compte.roles)]
